Вторая часть проектной работы

EldenLordCommander opened the pull request:
Здравствуйте! 
Вторая часть проектной работы:
- добавлено модельное окно с подложкой, закрывается на крестик, нажатие вне окна, клавишу Esc
- добавлен компонент Заказа
- Добавлен компонент просмотра ингредиента
- добавлен запрос к Апи

EldenLordCommander committed:
Вторая часть проектной работы

Добавлено модальное окно с подложкой, запросы к Апи при монтировании, два компонента - ингридиент и заказ

gennady-bars commented on src/components/app/app.tsx:
```diff
@@ -1,20 +1,43 @@
+               <BurgerConstructor data={state.data} />
+            </main >
+         }
+         <div id="modal"></div>
```
`<div id="modal"></div>` должно быть в `index.html` в папке `public`. Реакт не должен рисовать верстку для портала

gennady-bars commented on src/components/app/app.tsx:
```diff
@@ -1,20 +1,43 @@
+      fetch(url)
+         .then((response) => {
+            if (!response.ok) {
+              throw new Error();
```
`              throw new Error();
`

нужно осмысленную ошибку выкидывать


```js

              throw new Error('ответ от сервера не ok');

```

gennady-bars commented on src/components/app/app.tsx:
```diff
@@ -1,20 +1,43 @@
+         .then( 
+            result => setState({data: result.data, success: result.success})
+            )
+         .catch(error => alert("Ошибка запроса: " + error))
```
Отлично, что ловите возможные ошибки в конце каждого запроса к серверу

gennady-bars commented on src/components/app/app.tsx:
```diff
@@ -1,20 +1,43 @@
+            result => setState({data: result.data, success: result.success})
+            )
+         .catch(error => alert("Ошибка запроса: " + error))
+   }, [])
```
Отлично, что используете пустой массив зависимостей `[]`, чтобы `useEffect` сработал 1 раз

gennady-bars commented on src/components/ingredient-details/ingredient-details.js:
```diff
@@ -0,0 +1,46 @@
+    //console.log(ingreedient);
+    return (
+        <div>
+            <img src={ingredient.image_large} className={detailsStyle.priceBlock}></img>
```
Каждой картинке (тегу `img`) нужен информативный атрибут `alt` (например: имя картинки вставлять), так как при ошибке загрузки картинки на сайте подсветится **название** этой картинки (которое в `alt`). Еще это нужно для **плохо видящих** и `незрячих` людей - для них есть специальные средства, которые распознают текст в `alt` и проговаривают его. Подробнее тут https://habr.com/ru/post/341810/

gennady-bars commented on src/components/modal/modal.js:
```diff
@@ -0,0 +1,62 @@
+        }
+        window.addEventListener('keydown', close)
+
+        return () => window.removeEventListener('keydown', close)
```
правильно удаляется обработчик

gennady-bars commented on src/components/modal/modal.js:
```diff
@@ -0,0 +1,62 @@
+
+    useEffect(() => {
+        const close = (e) => {
+            if (e.keyCode === 27) {
```
Цифры лучше не использовать для определения клавиш, так как они не говорят ничего о себе (какая это клавиша). Лучше использовать строки. `(evt.key === “Escape”)`. И в чек-листе есть пункт про то, что нельзя использовать `магические` значения, которые не являются понятными любому разработчику.

gennady-bars commented on src/components/modal/modal.js:
```diff
@@ -0,0 +1,62 @@
+    }, [])
+
+    useEffect(() => {
+        const close = (e) => {
```
`close` нужно чуть более понятно назвать, чтобы отображало, что это обработка `Escape`


`handleEscape`

gennady-bars commented on src/components/modal/modal.js:
```diff
@@ -0,0 +1,62 @@
+
+    return ReactDOM.createPortal(
+        <>
+            <ModalOverlay closeModal={closeModal}></ModalOverlay>
```
Можно использовать **самозакрывающиеся** теги, если нет `children` (нет ничего между тегов). 


```jsx


      <ModalOverlay closeModal={closeModal} />

```

gennady-bars commented on src/components/modal/modal.js:
```diff
@@ -0,0 +1,62 @@
+            <div className={modalStyles.modal}>
+                <section>
+                    {
+                        isOrder === true
```
`isOrder `


модальное окно не должно знать про заказы. Это универсальный компонент для любых целей

gennady-bars replied:
это нужно проверять по `title`, в котором будет текст заголовка

gennady-bars commented on src/components/modal/modal.js:
```diff
@@ -0,0 +1,62 @@
+                            : (
+                                <div className={modalStyles.modalHeader}>
+                                    <div className={`${modalStyles.title} text text_type_main-medium`}>
+                                        Детали ингридиента
```
` Детали ингридиента`

внутри компонента `Modal` не должно быть **хардкода** строк и компонентов. 

Текст заголовка нужно передавать в виде `title` в пропсы

gennady-bars reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

Работа проделана огромная


* Отлично, что ловите возможные ошибки в конце каждого запроса к серверу
* Отлично, что используете пустой массив зависимостей `[]`, чтобы `useEffect` сработал 1 раз
* Хорошая типизация

но есть некоторые недочеты: 

* `<div id="modal"></div>` должно быть в `index.html` в папке `public`. Реакт не должен рисовать верстку для портала. В этом суть портала, чтобы отделиться от основного блока приложения
*  Скрин https://disk.yandex.ru/i/wppWKiHhz_qfSA Внутри заказа не должно быть **других булок**: есть только верхняя и нижняя одинаковые.
* Нужно подсчитать стоимость отображенного бургера. Скрин https://disk.yandex.ru/i/cOx-N1pkdOP5GQ 
* Каждой картинке (тегу `img`) нужен информативный атрибут `alt` 
* Цифры лучше не использовать для определения клавиш, так как они не говорят ничего о себе (какая это клавиша). Лучше использовать строки. `(evt.key === “Escape”)`. И в чек-листе есть пункт про то, что нельзя использовать `магические` значения, которые не являются понятными любому разработчику.
* модальное окно не должно знать про заказы. Это универсальный компонент для любых целей
* внутри компонента `Modal` не должно быть **хардкода** строк и компонентов. 


Исправьте, пожалуйста, недочеты и работа будет принята. Пожалуйста, проверьте **работоспособность проекта** и наличие возможных ошибок в консоли браузера (кнопка `F12`) перед отправкой на ревью. 

Напоминаю, что работа может быть принята только после исправления всех критических замечаний `Нужно исправить`.

Удачного рефакторинга кода.

EldenLordCommander committed:
Fix Ошибок по ревью

Поправил указанные моменты.
Вопрос только по Модалке - я убрал хардкор заголовка, убрал его в проп, но оставил условный блок, зависящий от наличия текста в title. Как я понял, такую конструкцию можно использовать

gennady-bars commented on src/components/modal/modal.js:
```diff
@@ -6,43 +6,41 @@ import { CloseIcon } from '@ya.praktikum/react-developer-burger-ui-components';
                     {
-                        isOrder === true
-                            ? ''
+                        title === '' ? ''
```
Да, Вы все правильно сделали


это универсально: есть заголовок? значит, добавляем верстку

gennady-bars reviewed (approved):
Поздравляю! Ваша работа принята.

Вы отлично потрудились.

Удачного дальнейшего обучения.
